### PR TITLE
Added proper support for merged non-runnable binaries

### PR DESCRIPTION
**Pull request by balavor**
## Merged non-runnable binaries support

**TL;DR**: Fixed frameworks and xcframeworks being embedded into the app after being merged to a non-runnable target (e.g., another dynamic framework).

### Short description 📝

Tuist supports mergeable frameworks introduced in Xcode 15.0 only for executables (apps, app extensions). The main purpose of tuist for mergeable support is to ensure that it doesn't embed merged frameworks separately into the app, preventing duplication.

However, if a separate framework is created to share the same dependencies between separate targets, Tuist will add merged local frameworks and xcframeworks to the list of embedded frameworks.

This MR fixes the issue for both frameworks and xcframeworks by inspecting the main target dependencies in the `embeddableFrameworks` phase of `GraphTraversal`. If any of them is a merged binary, its mergeable dependencies are excluded from the embedded list.

### How to test the changes locally 🧐

- Generate Project for `ios_app_with_shared_mergeable_framework` fixture
- Open Run Phases -> Embedded frameworks for the main App.
- Ensure only `SharedDependencies.framework` is embedded

### Results
| Before | After |
| ------ | ------ |
| image | image |

### Notes and potential improvements for the future

If a merged framework with mergeable dependencies is created, it has to be linked to the main targets instead of those dependencies that were merged. An extra linting phase could probably be added to ensure that behavior. 

### Contributor checklist ✅

- [x] The code has been linted using run `make workspace/lint-fix`
- [x] The change is tested via unit testing, acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 1]
```diff
@@ -266,27 +266,71 @@ public class GraphTraverser: GraphTraversing {
...
+        // Search for merged non-runnable binaries
+        for innerTarget in otherTargetFrameworks {
+            guard let graphTarget = self.target(from: innerTarget),
+                  graphTarget.target.mergedBinaryType != .disabled
+            else {
+                continue
+            }
+
+            // Take all mergeable dependencies
+            let mergeableFrameworks = filterDependencies(
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/GraphTraverser.swift:302]: I'd turn this one into a function and invoke it only in the case of `.automatic` and `.manual`. Otherwise we'll invoking it in the case of `.disabled`, even if we don't need it.

```swift
func mergeableFrameworks() -> [...] {
 ...
}
```

Comment (pepicrft) [reply Sources/TuistCore/Graph/GraphTraverser.swift:302]: Or even better, a private function in `GraphTraverser` to prevent adding more nesting to this function, which has already got too complex.

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 2]
```diff
@@ -266,27 +266,71 @@ public class GraphTraverser: GraphTraversing {
...
+                from: innerTarget,
+                test: isDependencyDynamicMergeableTarget,
+                skip: canDependencyEmbedProducts
+            )
+            switch graphTarget.target.mergedBinaryType {
+            case .automatic:
+                // Filter all dependencies from lists of precompiled binaries and local targets
+                precompiledFrameworks = precompiledFrameworks
```
Comment (pepicrft) [inline Sources/TuistCore/Graph/GraphTraverser.swift:310]: Shouldn't we filtering out from `precompiledFrameworks` direct and transitive pre-compiled dependencies of `innerTarget`? Or am I missing anything?

**Review by pepicrft — APPROVED** [review]
Thanks a lot for taking the lead on this one @balavor. Also, thanks for adding such a thorough set of tests covering different scenarios. The `GraphTraverser` is one of the most critical pieces of Tuist where regressions can happen easily so we'd better have it well covered with tests.
I'd wait for an additional look from @kwridan or @waltflanagan who have deeper experience on the traversing logic. They might be able to catch issues or suggest optimizations.

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 3]
```diff
@@ -266,27 +266,71 @@ public class GraphTraverser: GraphTraversing {
...
-        ) })
-
         /// Other targets' frameworks.
         var otherTargetFrameworks = filterDependencies(
             from: .target(name: name, path: path),
             test: isDependencyDynamicTarget,
             skip: canDependencyEmbedProducts
         )
 
+        // Skip merged precompiled libraries from merging into the runnable binary first
```
[the remark this answers is not in the document]
Comment (balavor) [reply Sources/TuistCore/Graph/GraphTraverser.swift:276]: If the binary is merged and the framework is mergeable, for Debug, it will be reexported, while for the Release, it will become a static framework, being a part of this merged binary.

So adding them into the embedded frameworks list makes no sense.

**Review by AlbGarciam — APPROVED** [review]

**Comment by pepicrft** [comment]
@balavor let me know when comments are addressed and if CI is happy we can go ahead with the merge. We've got a review by @AlbGarciam, who contributed the first iteration of that logic.

**Review by kwridan — COMMENTED** [review]
Thanks @balavor 

A few things I noticed while testing this

I'm not sure we have all the cases handled correctly for embedding here:

- Attempting to launch the sample app in the simulator results in the app extension crashing due to missing library
- Ignoring the app extension for a moment (removing it) and testing the app on its own without it, when attempting to build and launch for the device the app would also crash due to missing embedded library

```
dyld[2073]: Library not loaded: @rpath/DynamicFrameworkA.framework/DynamicFrameworkA
  Referenced from: <6725BAE8-2E6C-3208-AE58-EADC354500CE> /private/var/containers/Bundle/Application/3377AD25-4B65-47E5-9EC1-C46D27FB0029/App.app/App
``` 

From my limited understanding of how mergeable libraries work, they are limited to targets within the same project only 🤔

xref: https://github.com/tuist/tuist/pull/5351

**Comment by balavor** [comment]
> @balavor, let me know when comments are addressed, and if CI is happy, we can go ahead with the merge. We've got a review by @AlbGarciam, who contributed the first iteration of that logic.

I'll do my best to address them by the EOF.

File: fixtures/ios_app_with_shared_mergeable_framework/Project.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    targets: [
+        Target(
+            name: "App",
+            destinations: .iOS,
+            product: .app,
+            bundleId: "io.tuist.App",
+            infoPlist: "App/Info.plist",
+            sources: ["App/Sources/**"],
+            dependencies: [
+                .project(target: "SharedDependenciesFramework", path: "Modules/SharedDependenciesFramework"),
```
[the remark this answers is not in the document]
Comment (balavor) [reply fixtures/ios_app_with_shared_mergeable_framework/Project.swift:14]: Hey @kwridan ,

Even though I suggest that this is an Apple's bug that mergeable only works in a single project, you are right. I'd change my example to avoid creating multiple **xcodeproj** files.

> Seeing the app directly imports DynamicFrameworkA and DynamicFrameworkB would we not need to be explicit about depending on them too? - I can see the SharedDependenciesFramework combines those.

Based on how it works, we only have to link and embed `SharedDependenciesFramework`. However, we should still use `import DynamicFrameworkA` and `import DynamicFrameworkB` in our codebase.

That's the point of this shared framework - it wraps all your dependencies and includes them in your code base either statically or dynamically, so you don't have to every time go into all of your targets and add your newly created frameworks (unless they are not mergeable).

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 2 above)
Comment (balavor) [reply Sources/TuistCore/Graph/GraphTraverser.swift:310]: Thanks, @pepicrft. That's an interesting question. From what I understand, if you have the following setup

```
- App 
    -> SharedFramework
- SharedFramework, dynamic
    -> frameworkA: dynamic, mergeable
    -> xcframeworkB: dynamic, mergeable
    -> xcframeworkC: dynamic, mergeable
```
you have 3 scenarios:

1. `SharedFramework` is not a merged binary
    - `frameworkA`, `xcframeworkB` and `xcframeworkC` should be embedded into the app
2. `SharedFramework` is a merged binary, marked as `automatic`
    - `frameworkA`, `xcframeworkB` and `xcframeworkC` should not be embedded into the app, because they are merged into the `SharedFramework`
4. `SharedFramework` is a merged binary, marked as `manual("xcframeworkC")`
    - `frameworkA`, `xcframeworkB` should be embedded into the app, while `xcframeworkC` - not, because it was defined that it should be merged.
    
Please correct me if you have any other knowledge on that, but that's what came up from all the docs and manipulations that I had with mergeable frameworks :)

Comment (balavor) [reply Sources/TuistCore/Graph/GraphTraverser.swift:310]: Overall, it is possible that the Project is defined in a following way:

```
- App 
    -> SharedFramework
    -> xcframeworkB
- SharedFramework, dynamic
    -> frameworkA: dynamic, mergeable
    -> xcframeworkB: dynamic, mergeable
    -> xcframeworkC: dynamic, mergeable
```

but such approach is **invalid** - the app should only link the `SharedFramework`, otherwise, `xcframeworkB` should be removed from the `SharedFramework`. 

There has to be a linter for such a thing, but I didn't have time yet to figure out how linters work in Tuist. So if you have any ideas how to easily write one, I might be able to open another MR.

**Review by AlbGarciam — APPROVED** [review]

**Review by kwridan — COMMENTED** [review]
Thanks @balavor 

I did some more digging into mergeable libraries and seems removing the embedded frameworks may not be required.

I'm curious, aside from the frameworks appearing in the embed phase, were there any build or runtime issues witnessed that motivated this change?

One key thing to test is introducing resources into the mix, and testing on device in both Debug and Release configurations.

For example:

- Make a project with the following graph
  - `App` > `FrameworkA` > `FrameworkB`
  - App: Application
  - FrameworkA: dynamic framework
  - FrameworkB: dynamic framework
- Mark `FrameworkB` as mergeable
- Mark `FrameworkA` with merged binary `.automatic`
- Add resources to `FrameworkA` and `FrameworkB`
- Attempt to access said resources (e.g. `Bundle(for: FrameworkBClass.self)`)

Notice without embedding both `FrameworkA` and `FrameworkB` in the app (regardless of mergeable state) the app won't be able to load the resources.

The WWDC talk does indeed mention with merged libraries only the one merged binary needs to be embedded 

>  We can create a framework that merges the contents of the other libraries. So you end up with only one framework to embed.[^1]

However this doesn’t seem to work when resources are included. Interestingly throughout the talk (possibly I missed it) the merged frameworks only got removed from the link build phase as opposed to the embed build phase.

The talk does go on to mention a “hook” to make resource look ups in bundles work with mergeable libraries 

> Similarly, resource lookup could be impacted by library merging. This is because of what the runtime expects. In Swift, bundle is an API to have the runtime load a framework's bundle. The equivalent API in Objective-C is NSBundle's bundleForClass. These APIs are used to work with a framework's resources without having to consider the bundle's structure.
> …
> Good news! In iOS 12, a hook was added to enable lookup for this scenario.[^2]

Searching around for what this hook is, there’s an Apple Forum post [^3]  as well as a great blog post [^4] that sheds some light on it. While all the frameworks are explicitly embedded in the app target, it seems the build system isn't duplicating the binaries, it has a binary with no symbols but retain the framework structure to support resources.

We can use the `nm` command to inspect which symbols are present in a binary:

```sh
nm /path/to/App.app/Frameworks/FrameworkB.framework/FrameworkB
...
no symbols
```

They would instead reside in `FrameworkA` 
- debug: within the re-exported `/path/to/App.app/Frameworks/FrameworkA/ReexportedBinaries/FrameworkB.framework/FrameworkB`
- release: merged into the `FrameworkA` binary

So it seems the build system is automagically handling this for us. 

Given all this, I’m not sure it’s necessary to remove frameworks from the embed phase 🤔 - have I missed some other pieces to this puzzle?

Thanks again @balavor 

[^1]: https://developer.apple.com/videos/play/wwdc2023/10268/?time=435
[^2]: https://developer.apple.com/videos/play/wwdc2023/10268/?time=1267
[^3]: https://developer.apple.com/forums/thread/736404
[^4]: https://medium.com/@SanjuNaik14/mergeable-libraries-with-cocoapods-2ad902a7ecec

**Comment by balavor** [comment]
Hey @kwridan,

Thanks for the deep research. That is really helpful. I might be trying to solve a different problem with those changes. Adding 35kb from local dependencies is not a huge tragedy since we don't have to manage those framework resources by ourselves.

> So it seems the build system is automagically handling this for us.

That is not true for  **xcframeworks**. For the experiment, I ran `Archive` for the example app, having 10 mergeable xcframeworks that are about 20Mb altogether. And what happens after that - all of their symbols are merged in a merged binary, **but**, all embedded frameworks from these xcframeworks are **not stripped**, so I'm getting an **extra 20Mb** of binary size in my app. 

A weird part is that I don't have a duplicated symbols issue, but that's how it works now. I guess the problem above is mentioned in a `Caveats` section of the article that you've shared.

If that's an Apple bug, we could try and wait for them to solve it, but it basically prevents using Mergeable for everyone who has a lot of xcframeworks. For example, we pre-build all our 3rd party dependencies in xcframeworks on our own, preventing redundant compilation in run time. 

Maybe we could add some script that will remove all resources from pre-built frameworks after it is merged?

**Comment by pepicrft** [comment]
> If that's an Apple bug, we could try and wait for them to solve it, but it basically prevents using Mergeable for everyone who has a lot of xcframeworks. For example, we pre-build all our 3rd party dependencies in xcframeworks on our own, preventing redundant compilation in run time.

I'll let @kwridan make the best call here, but after reading the state of things, I'm inclined to have something temporary until Apple addresses the issue on their end. What would the strip command look like?

**Comment by kwridan** [comment]
Thanks for the additional details @balavor 

> That is not true for xcframeworks. For the experiment, I ran Archive for the example app, having 10 mergeable xcframeworks that are about 20Mb altogether. And what happens after that - all of their symbols are merged in a merged binary, but, all embedded frameworks from these xcframeworks are not stripped, so I'm getting an extra 20Mb of binary size in my app.

Ouch, this doesn't sound great 😞 

>  I'm inclined to have something temporary until Apple addresses the issue on their end. What would the strip command look like?

It's worth filing an apple feedback either way @balavor. Curious which command we can use to remove the symbols after the XCFramework is embed in the app bundle.

My worry is there may not be a safe script to run for all cases / projects. I'm thinking perhaps we can limit the logic of omitting the products from the embed phase for merged xcframeworks only? The main downside here is, xcframeworks with resources will not work correctly.

**Comment by balavor** [comment]
Closing this MR. In Xcode 15.3, Apple fixed the issue by properly stripping both xcframeworks and frameworks from another project.
